p2p/router.go

```diff
@@ -0,0 +1,568 @@
+						"peer", envelope.To, "channel", channel.id)
+					continue
+				}
+
+				select {
+				case peerQueue.enqueue() <- envelope:
+				case <-peerQueue.closed():
+					r.logger.Error("dropping message for non-connected peer",
+						"peer", envelope.To, "channel", channel.id)
+				case <-r.stopCh:
+					return
+				}
+			}
+
+		case peerError, ok := <-channel.errCh:
+			if !ok {
+				return
+			}
+			// FIXME: We just disconnect the peer for now
+			r.logger.Error("peer error, disconnecting", "peer", peerError.PeerID, "err", peerError.Err)
+			r.peerMtx.RLock()
+			peerQueue, ok := r.peerQueues[peerError.PeerID]
+			r.peerMtx.RUnlock()
+			if ok {
+				peerQueue.close()
+			}
+
+		case <-channel.Done():
+			return
+		case <-r.stopCh:
+			return
+		}
+	}
+}
+
+// acceptPeers accepts inbound connections from peers on the given transport.
+func (r *Router) acceptPeers(transport Transport) {
+	for {
+		select {
+		case <-r.stopCh:
+			return
+		default:
+		}
+
+		conn, err := transport.Accept(context.Background())
+		switch err {
+		case nil:
+		case ErrTransportClosed{}, io.EOF:
+			r.logger.Info("transport closed; stopping accept routine", "transport", transport)
+			return
+		default:
+			r.logger.Error("failed to accept connection", "transport", transport, "err", err)
+			continue
+		}
+
+		peerID := conn.NodeInfo().NodeID
+		if r.store.Claim(peerID) == nil {
+			r.logger.Error("already connected to peer, rejecting connection", "peer", peerID)
+			_ = conn.Close()
+			continue
+		}
+
+		queue := newFIFOQueue()
+		r.peerMtx.Lock()
+		r.peerQueues[peerID] = queue
+		r.peerMtx.Unlock()
+
+		go func() {
+			defer func() {
+				r.peerMtx.Lock()
+				delete(r.peerQueues, peerID)
+				r.peerMtx.Unlock()
+				queue.close()
+				_ = conn.Close()
+				r.store.Return(peerID)
+			}()
+
+			r.routePeer(peerID, conn, queue)
+		}()
+	}
+}
+
+// dialPeers maintains outbound connections to peers.
+func (r *Router) dialPeers() {
+	for {
+		select {
+		case <-r.stopCh:
+			return
+		default:
+		}
+
+		peer := r.store.Dispense()
+		if peer == nil {
+			r.logger.Debug("no eligible peers, sleeping")
+			select {
+			case <-time.After(time.Second):
+				continue
+			case <-r.stopCh:
+				return
+			}
+		}
+
+		go func() {
+			defer r.store.Return(peer.ID)
+			conn, err := r.dialPeer(peer)
+			if err != nil {
+				r.logger.Error("failed to dial peer, will retry", "peer", peer.ID)
+				return
+			}
+			defer conn.Close()
+
+			queue := newFIFOQueue()
+			defer queue.close()
+			r.peerMtx.Lock()
+			r.peerQueues[peer.ID] = queue
+			r.peerMtx.Unlock()
+
+			defer func() {
+				r.peerMtx.Lock()
+				delete(r.peerQueues, peer.ID)
+				r.peerMtx.Unlock()
+			}()
+
+			r.routePeer(peer.ID, conn, queue)
+		}()
+	}
+}
+
+// dialPeer attempts to connect to a peer.
+func (r *Router) dialPeer(peer *peerInfo) (Connection, error) {
+	ctx := context.Background()
+
+	for _, address := range peer.Addresses {
+		resolveCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
```

Comment:
should the timeout be configurable?

Reply:
Probably, eventually. This is just a prototype, we'll be adding/integrating configurability later.